### ISSUE TITLE
SVD backward isn't correctly defined when multiple 0 eigenvalues

### ISSUE THREAD

**Issue by AdrienCourtois** [pytorch/pytorch#49886]
## 🐛 Bug

SVD backward function doesn't deal properly with the case where there are multiple zero eigenvalues. It seems like the term 1/(\sigma_i - \sigma_j) is computed even when \sigma_i = \sigma_j = 0. 

Therefore, all non-invertible matrices A with dim(Ker(A)) > 1 cannot be back-propagated through.

The work [https://j-towns.github.io/papers/svd-derivative.pdf](https://j-towns.github.io/papers/svd-derivative.pdf) seems to indicate that it is possible to derive a solution for the gradient that takes into account this case.

## To Reproduce

```python
import torch
import torch.nn.functional as F

random_init = False

x = torch.zeros(15,3)
if random_init:
    x.normal_()
else:
    for i in range(x.size(1)):
        x[i,i] = i+1

# Generate rank-3 matrix
y = torch.zeros(15, 15)
y[:,:5] = x[:,0][:,None]
y[:,5:10] = x[:,1][:,None]
y[:,10:] = x[:,2][:,None]

y.requires_grad_(True)

U, S, V = torch.svd(y)
print(S)

z =  U @ torch.diag_embed(S) @ V

loss = z.sum()
loss.backward()

torch.all(y.grad == y.grad)
```

```
> tensor([6.7082, 4.4721, 2.2361, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000], grad_fn=<SvdBackward>)
> tensor(False)
```

## Expected behavior

The gradient should not have any `nan` values.

## Environment

I am using the default environment provided by Google Colab.

PyTorch version: 1.7.0+cu101
Is debug build: True
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.3.1
[pip3] torchvision==0.8.1+cu101
[conda] Could not collect

## Additional context

You can also note that the eigenvalues are subject to slight approximation errors (for instance you can note that if you let `random_init` to be `True`), which results in zero eigenvalues being approximated as very small floats. This should also be taken into account when computing the gradient.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @jlin27 @mruberry @jianyuh @heitorschueroff @walterddr

**Comment by nikitaved** [comment]
As far as I know, it is indeed possible to implement SVD/SVD_backward that is rank-aware, but for that we need to explicitly know the value of rank which in turn requires synchronization with the CPU if Tensors are on GPU.
It is impossible to avoid this synchronization for rank-deficient inputs, unfortunately.

For now it is probably worth updating the documentation that backward for SVD is only implemented for full-rank inputs.

**Comment by albanD** [comment]
High pri to update the doc for now

**Comment by ngimel** [comment]
Related: #47599

**Comment by gchanan** [comment]
the symeig backward note has been added to 1.8.0 so I'm removing the milestone label.

**Comment by vadimkantorov** [comment]
Also, maybe this approach could be adopted: https://github.com/WeiWangTrento/Robust-Differentiable-SVD, https://arxiv.org/abs/2104.03821

**Comment by dagitses** [comment]
> High pri to update the doc for now

Per-offline discussion, @albanD believes the doc already covers this case. From `torch.linalg.svd`:

> Gradients computed using U or Vh will only be finite when A does not have zero as a singular value or repeated singular values.

**Comment by mruberry** [comment]
cc @Lezcano @nikitaved -- maybe we can close this now?

**Comment by nikitaved** [comment]
The problem is not yet solved and it does persist not just for singular inputs or rank `r < n -1` but also for full-rank inputs with repeated singular values.

**Comment by Lezcano** [comment]
The SVD map is not differentiable when there are repeated singular values or zero singular values.
When there is a zero singular value, `svdvals` gives the right gradients, but you have problems computing the gradients for `U` or `V` as you need to compute the inverse of `S`, as the docs suggest:
https://github.com/pytorch/pytorch/blob/056a8e0d5cc6b53dd574b77eebf273ded646f81e/torch/csrc/autograd/FunctionsManual.cpp#L2132

That paper is making a number of assumptions implicitly. We got similar proposals [here](https://github.com/pytorch/pytorch/pull/50942) for `eigh` and I investigated with @nikitaved similar formulas for the QR. The general problem is described geometrically [in this post](https://github.com/pytorch/pytorch/issues/47599#issuecomment-812539850).

Now, on what to do. We discussed that a reasonable way to go about this would be:
1) Implement the formula where the function is differentiable
2) Document where each function is differentiable and where it is not
3) Figure out whether there exists a framework for which the formulae that stay in the low-dimensional manifold make sense
3.1) Provided we can formalise this, implement these formulae when possible as long as they do not make the backwards much slower in the general case.

At the moment we have implemented 1) and 2) for `torch.linalg`, but 3) is a big one. As far as I know, there is no document that defines formally what is a subgradient for a matrix function and proves that the subgradients for these functions take certain form. There is this issue open to discuss these things: https://github.com/pytorch/pytorch/issues/57272
I'd say we can close this one for now (and related issues like https://github.com/pytorch/pytorch/issues/47599, and if anyone has any pointers on how to go make progress on 3) they could share them in https://github.com/pytorch/pytorch/issues/57272.

**Comment by Lezcano** [comment]
This behaviour is now documented in `torch.linalg.svd` and is expected.

Closing this issue in favour of https://github.com/pytorch/pytorch/issues/57272 which tracks all these differentiability issues for matrix functions.
